Fix ordering of invocation on Resources

rvansa opened the pull request:
* keeps the original handling of exceptions: afterRestore is called even if beforeCheckpoint throws
* allows to register a resource in a context that did not start beforeCheckpoint invocations yet
* registering resource in previous/running context fails the checkpoint but does not trigger exception immediately
   * instead this will be one of the recorded exceptions and the resource has a chance to fire next time
* allowed registration of resources can be invoked from other thread without deadlock; illegal registration can deadlock, though

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace

### Reviewers
 * [Anton Kozlov](https://openjdk.org/census#akozlov) (@AntonKozlov - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/crac.git pull/60/head:pull/60` \
`$ git checkout pull/60`

Update a local copy of the PR: \
`$ git checkout pull/60` \
`$ git pull https://git.openjdk.org/crac.git pull/60/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 60`

View PR using the GUI difftool: \
`$ git pr show -t 60`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/crac/pull/60.diff">https://git.openjdk.org/crac/pull/60.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/crac/pull/60#issuecomment-1517536895)

AntonKozlov commented on src/java.base/share/classes/jdk/crac/impl/AbstractContextImpl.java:
```diff
@@ -22,86 +22,178 @@
+                    r.beforeCheckpoint(this);
+                    restoreQ.add(r);
+                } catch (CheckpointException e) {
+                    enqueueIfContext(r);
```
Ohh, I see the problem

> When Context.beforeCheckpoint throws, invoke Context.afterRestore anyway (otherwise some resources stay in suspended state).

So this is proposed to be handled in the parent context? Have you considered fixing that in the child context, run afterRestore for successfully checkpointed resources, before throwing CheckpointException to the parent context?

rvansa replied:
Right now we run the checkpoint for all resources even if some fail. Are you fine with erroring out on the first exception?

AntonKozlov replied:
Why is that necessary? If we are trying to avoid second afterRestore for sucessfully checkpointed Resources, that would mean that Context.afterRestore is called, that controverts the problem definition.

Assuming afterRestore is not called for the Resource with failed beforeCheckpoint. Suppose we have R1, R2 in a child Context, and R1.beforeCheckpoint succeed, and R2.beforeCheckpoint fails. We can call R1.afterRestore, before throwing from Ctx.beforeCheckpoint. That will restore successfully checkpointed resources.

But I checked the code and it seems that we do call afterRestore regardless of the result of beforeCheckpoint 

https://github.com/openjdk/crac/blob/95394e84683f1a816c0283f8c834072324516fba/src/java.base/share/classes/jdk/crac/impl/AbstractContextImpl.java#L54

Every Resource available is collected in List resources and then restore queue is exactly the reverse of that list, it does not matter if any of resources throwed CheckpointException. Right?

rvansa replied:
On regular resources `afterRestore` is not called if the resource throws, it does not end up in the restoreQ.

I am a bit worried about calling `R3.bC` that would live in a different context when `R1` is already restored. The components were called in certain order for a reason, and if R3 expect R1 to be checkpointed this could end up badly.
You might argue that R2 is not in checkpointed state anyway and R3 might expect it to be, but at least here one component failed and there is a track of consequences. On the other hand R1 and R3 might be completely unrelevant to R2 and the fact that R2 failed shouldn't interfere.

What is the point of continuing with the checkpoint when we now that it won't happen in the end, with all the suppressed exceptions in the first place? Just to collect all errors at once rather than one-by-one? Is it worth potential false positives?

AntonKozlov replied:
> On regular resources afterRestore is not called if the resource throws, it does not end up in the restoreQ.

I'm puzzled how this is possible in the implementation before the patch. The code says it should
```
        List<R> resources = checkpointQ. ... ;
        for (Resource r : resources) { ... }
        Collections.reverse(resources);
        restoreQ = resources;
```

We'll have to document the expected behavior, but that is another concern.

> I am a bit worried about calling R3.bC that would live in a different context when R1 is already restored.

This is a question of specification. If we set that, it will be adopted by rearranging Contextes and Resources, pretty manageable.

In general, resources should not have a lot of assumptions about other resources. Otherwise, they should directly refer to each other, rathen than relying on the Context. That will be more straightforward and less error-prone.

> What is the point of continuing with the checkpoint when we now that it won't happen in the end, with all the suppressed exceptions in the first place? Just to collect all errors at once rather than one-by-one? Is it worth potential false positives?

Correct, to run all registered code, that may also have some side-effects (unpreferable, but possible). Having that checkpoint won't happen anyway, false positive does not look a problem.

rvansa replied:
True, checking again before the patch it was really calling it on resources even after failure - I had "fixed" that probably way before and remembered it that way. When a code is throwing it is aware of that, so it should either revert automatically or be considered completely broken. I would bring an analogy with a locking code - `lock()` method is executed outside of the try-finally block so if it fails it should keep the lock unlocked.

The R1 and R3 issue - it is fixable, true, but it's the infra code that introduced the failure. User sees both R2 and R3 failures, decides to investigate R3 only to find that it was a red herring.

> resources should not have a lot of assumptions about other resources

JDK code is a direct contradiction to that. We shall not expect user code to be any different.

> false positive does not look a problem

Red herrings are UX problem. And it is a problem when the failure leaves the JVM in a state that does not allow inspection - not sure why but I had cases where neither `jstack` nor `kill -3` worked.

AntonKozlov commented on src/java.base/share/classes/jdk/crac/impl/AbstractContextImpl.java:
```diff
@@ -22,86 +22,178 @@
+            }
+            locked = true;
+            // This is important for the case of recursive registration
+            throwIfCheckpointInProgress(priority);
```
What is expected handling of this exception? And in the current form the exception is not checked, so in most cases that exception won't be expected. Having that in most cases resources are registered during class or object initialization, those entities will be caught in partially constructed states, likely leaving the whole system stuck.

Would it not better to allow registration, but throw CheckpointException at the end of Context.beforeCheckpoint? That is pretty legit exception that means the checkpoint can be attempted again, and another attempt can be pretty well successful, if all Resources do not throw and no new Resources are registered.

rvansa replied:
So, making `register` throw a checked exception wouldn't be a good solution; most of the time there wouldn't be a better handling than to log && rethrow.

What you propose - having a form of exception stack in the context and checking it anything was added after `beforeCheckpoint` could work, though this is definitely not a pattern that would be natural in Java - it smells of JNI developer :)

I think that a situation when the ISE is thrown is a programming error, rather than reacting on the state of 'outer world' (as in case of e.g. I/O...) and as such this can stay as unchecked exception, possibly leading to inconsistent state of the program.

rvansa replied:
Generally speaking, your proposal to try performing the checkpoint until it succeeds is more permissive, as users can back-off from a bad/fragile design through retrying, rather than being forced to correct the dependencies between components. In the past it seemed to me that you were more on the "good design required" rather than "practical" side of things - please don't take this as any form of "ad hominem" argument, I would just like to understand why you opt for such solution here. In my view a timeout/IO error are a totally different category of error.

AntonKozlov replied:
The CheckpointException is a CRaC-specific, deliberately thrown exception to communicate the problem with the state to the requester of the checkpoint. The exception may suggest rewriting the code, or another attempt. We don't care a lot about the preceise meaning, we just need to deliver the exception. Although lesser number of exceptions is better of course.

rvansa replied:
I realized that you actually asked for a different behaviour than what I thought initially, something like the `ConcurrentModificationException` while iterating a collection. That is not that bad, and we can add debug option to track where this happened.

rvansa replied:
Thinking about this again, the current code does not prevent that from happening either, had JDK context registered something new in the global context this would be silently ignored.
I can imagine that Core could keep a checkpoint counter and the Context would record last finished counter. During registration, if the current counter equals finished counter the registration could throw. However if we're to rollback the checkpoint, we would probably need to set a global flag in Core, too.

AntonKozlov replied:
Yes, the current implementation allows new Resources registered during notification and silently ignored. I hope this change will deal with the problem somehow.

> The best you could do is to invoke the beforeCheckpoint on the resource when you find C1 to be done, but this is not what we discussed before.

This is an interesting approach, but this won't suit e.g. Global Context, as last registered Resource would need to be notified first. If notification is in progress, we've lost our chance.

But I assume some Context implementation may opt to this, so we need to leave some room in the spec to allow such implementations.

rvansa commented on src/java.base/share/classes/jdk/crac/Resource.java:
```diff
@@ -37,6 +37,8 @@ public interface Resource {
 
     /**
      * Invoked by a {@code Context} as a notification about checkpoint.
+     * Order of checkpoint notification is the reverse order of
+     * {@link Context#register(Resource) registration}.
```
Alright, I think that the order for the global context is rather hidden in the `package-info.java`. So, I'll add a note here that the order is defined by the context, and move this info to `Core.getGlobalContext` javadoc.

Can we say that the fact that afterRestore is called in an inverse order to beforeCheckpoint is an universal rule, or would you keep that specific to context impl?

rvansa commented:
@AntonKozlov It took some effort but I managed to implement the semantics you have requested.

rvansa commented on src/java.base/share/classes/jdk/crac/impl/PriorityContext.java:
```diff
@@ -0,0 +1,101 @@
+// Copyright 2023 Azul Systems, Inc.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+//
+// 1. Redistributions of source code must retain the above copyright notice,
+// this list of conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright notice,
+// this list of conditions and the following disclaimer in the documentation
+// and/or other materials provided with the distribution.
+//
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
```
Uh, OK - I've copy pasted what's in the AbstractContextImpl as I've forked the class...

rvansa commented on src/java.base/share/classes/jdk/crac/impl/AbstractContextImpl.java:
```diff
@@ -22,86 +22,85 @@
+        restoreQ.add(resource);
+        try {
+            resource.beforeCheckpoint(semanticContext());
+        } catch (CheckpointException e) {
+            recordExceptions(e);
+        } catch (Exception e) {
+            Core.recordException(e);
```
The contexts are supposed to run all resources regardless of whether any failed, therefore there's no point in propagating and re-wrapping the exceptions. We could have a method to check whether this C/R is 'marked for rollback' (has any exceptions), but I don't see why would the parent context decide on the type - because it would get just CheckpointException with a list of other failures from components it does not understand.

About the distinction: the difference is that if you get a CheckpointException you'd unwrap it, recording only the inner suppressed ones. But I should push that to `recordExceptions` and rather decide based on CheckpointException message than number of suppressions.

AntonKozlov replied:
> The contexts are supposed to run all resources regardless of whether any failed, therefore there's no point in propagating and re-wrapping the exceptions. ...  I don't see why would the parent context decide on the type - because it would get just CheckpointException with a list of other failures from components it does not understand.

The parent Context may implement an artbitrary handling (for example, unloading a component compleltely if that is throwing an exception). So throwing an exception is something useful.

With that, the new Core.recordException is completely new exception flow, that just opimizes somthing the generic throw scheme. With that, the generic schemes should be something good enough already, we don't need to complicate the interface, the code,.. CheckpointExcepotions are still exceptions, that is, we don't expect them often, there is no need to optimize them.

rvansa replied:
It's not about optimization (in the sense of performance) but about removing code bloat, and the need for parent context to tediously copy failures (deciding whether something was a wrapper exception or the actual failure), when all you need to do in the end is to report them in bulk. It's not a new exception flow, it's removing the flow as there is no exception flow needed.

Your example is invalid: If a throwing resource is to be removed, it's the task of the parent context - and that one will see the exception. The parent context should not remove its child context since one of the N resources in the child context is failing.

rvansa replied:
> You're trading additional code for a complicated interface. What are the directions what to use: Core.recordException() vs throw new Exception() ? This two very similar interfaces are the sign we are trying to do something strange.

That's because of the non-standard requirement to continue after encountering an exception, and multiple rewraps in a hierarchy of contexts.
Resources should normally throw exceptions; Contexts are here to call into the Resource and aggregate errors. There's no point of propagating error higher up the hierarchy.

> Could you elaborate?.. I'm lost what "remove child context" means. I was reffering to the parent context being able to handle a CheckpointException from the child context.

And why would the child context throw the CheckpointException? The most common reason would be that one of X resources in child context has thrown. It's fine if a (child) Context removes the inner Resource after throwing. But that error should not propagate any higher, because if parent context were to remove it's (throwing) child context it would remove it along with the other X - 1 resources that are working correctly.

rvansa replied:
I realized I haven't stressed enough one motivation for `Core.registerException`: when the context has finished its `beforeCheckpoint` (it won't be called again) and someone calls into this Context's `register` we cannot propagate the failure through throwing. So we won't avoid API change - and in my view it's natural to use the same method for reporting the failure during `beforeCheckpoint` execution, too.

rvansa commented on src/java.base/share/classes/jdk/crac/impl/AbstractContextImpl.java:
```diff
@@ -22,86 +22,85 @@
 
-        Collections.reverse(resources);
-        restoreQ = resources;
+    protected abstract void runBeforeCheckpoint();
```
ACI is implemented both by OrderedContext and PriorityContext, while PC is quite different from OC.

AntonKozlov replied:
I'm trying to describe class hierachy and failing. The patch tries to reverse ACI-PC relation. ACI was for partially-ordered Resources (defined by a Comparator), and now it's for totally-ordered Resources (ordered by long). Trying to fit the partially-ordering PC as a subclass of the totally-ordering ACI feels unnatural. Can we have a cleaner hierachy of the classes?

rvansa replied:
No; ACI was totally-ordered in the previous version of the PR, but now it doesn't care about ordering at all; it's the abstract `runBeforeCheckpoint` that decides on the order.

rvansa replied:
We can further fine-tune the separation when the need arises and we have a concrete example.

rvansa commented on src/java.base/share/classes/jdk/crac/impl/AbstractContextImpl.java:
```diff
@@ -22,86 +22,85 @@
+        // Resource.afterRestore is invoked even if Resource.beforeCheckpoint fails
+        restoreQ.add(resource);
+        try {
+            resource.beforeCheckpoint(semanticContext());
```
Theoretically, the method could do that. However, here the purpose of `semanticContext()` is to pass the context to which it was registered rather than the subcontext where it is stored (but this is an implementation detail that the resource does not know about).

AntonKozlov replied:
Just realized that this is required for PriorityContext implementation. But that is the implementation of that class, it's wrong ACI has to care about that.

rvansa replied:
Yes, it's a bit of enforced flexibility of the base class (through allowing to override a method), though it doesn't need about the use case.

I could use just a list rather than sub-contexts but that would require duplicated code.

rvansa commented on src/java.base/share/classes/jdk/crac/impl/OrderedContext.java:
```diff
@@ -26,28 +26,62 @@
+    @Override
+    public synchronized void register(R r) {
+        resources.put(r, order++);
+        // It is possible that something registers to us during restore but before
+        // this context's afterRestore was called.
+        if (checkpointing && !Core.isRestoring()) {
```
Does that matter, though? The registration itself will be done for next C/R in all cases. The situation where the registrar entered the `register()` method, added the resource and recorded an exception but this has been silently discarded is equivalent to the situation where it entered the `register()` just after restore was completed. If these are independent you cannot force it to fail reliably, unless there is something in running the resource notifications that depends on the registration (but it's not as all resources have been notified).

If the intend was to do something after the current restore the resource should have been registered in a blocking way before the checkpoint or as part of the blocking beforeCheckpoint, not absolutely independently.

AntonKozlov replied:
> Does that matter, though? The registration itself will be done for next C/R in all cases. The situation where the registrar entered the `register()` method, added the resource and recorded an exception but this has been silently discarded is equivalent to the situation where it entered the `register()` just after restore was completed. 

Suppose aR() has some really important side-effect, it's totally necessarily to run that on restore. Then it falls to the category of problems this PR tries to solve (silently ignoring registered resources). 

> If these are independent you cannot force it to fail reliably, unless there is something in running the resource notifications that depends on the registration (but it's not as all resources have been notified).
>
> If the intend was to do something after the current restore the resource should have been registered in a blocking way before the checkpoint or as part of the blocking beforeCheckpoint, not absolutely independently.

Some form of the guarantee will be good. The blocking registration, I assume it something like register() to finish only when the argument is successfully registered? This looks like a viable approach, e.g. it does specify the behavior around possible race and do not affect "normal" workflow when registration happens way before the checkpoint. Do you see any problem with blocking registration?

rvansa commented on src/java.base/share/classes/jdk/crac/Core.java:
```diff
@@ -76,56 +80,60 @@ private static class FlagsHolder {
 
     /**
      * Gets the global {@code Context} for checkpoint/restore notifications.
+     * Order of invoking {@link Resource#beforeCheckpoint(Context)} is the reverse
+     * of the order of {@link Context#register(Resource) registration}.
+     * Order of invoking {@link Resource#afterRestore(Context)} is the reverse
+     * of the order of {@link Resource#beforeCheckpoint(Context) checkpoint notification},
+     * hence the same as the order of {@link Context#register(Resource) registration}.
```
Okay, I can remove the description in package. As for `javax.crac` - I thought that this should be really a mirror of `jdk.crac`, why the distinction?

Another way might be to make OrderedContext a marker interface (move the implementation to OrderedContextImpl) and put the description there, using this interface for `getGlobalContext()`.

rvansa commented on src/java.base/share/classes/jdk/crac/impl/AbstractContextImpl.java:
```diff
@@ -22,86 +22,85 @@
-    private static class FlagsHolder {
-        public static final boolean DEBUG =
-            GetBooleanAction.privilegedGetProperty("jdk.crac.debug");
+    protected static <E extends Exception> void recordExceptions(E source) {
+        Throwable[] suppressed = source.getSuppressed();
+        if (suppressed.length == 0) {
+            Core.recordException(source);
+        }
+        for (Throwable t : suppressed) {
+            Core.recordException(t);
+        }
```
This is how it's actually used; can't use union type in Java. I'll add an assertion to make this clear...

rvansa commented on src/java.base/share/classes/jdk/crac/impl/AbstractContextImpl.java:
```diff
@@ -22,86 +22,85 @@
+        // beforeCheckpoint and afterRestore should be invoked only
+        // by the single thread performing the C/R and other threads should
+        // not touch that.
+        restoreQ = new ArrayList<>();
+        runBeforeCheckpoint();
+        Collections.reverse(restoreQ);
```
Not really; the task of ACI subclass (OC, PC...) is to call `invokeBeforeCheckpoint` on some resources (ACI does not know which ones) in some order. The task of ACI is to remember the order of invocations and in `afterRestore` call this in a reversed order; the subclass does not need to know about any collection used for that.

rvansa commented on src/java.base/share/classes/javax/crac/package-info.java:
```diff
@@ -67,29 +67,6 @@
- *   </li>
- *   <li>When an exception is thrown during notificaion, it is caught by the {@code Context} and is suppressed by a {@code CheckpointException} or {@code RestoreException}, depends on the throwing method.
- *   </li>
- *   <li>When the {@code Resource} is a {@code Context} and it throws {@code CheckpointException} or {@code RestoreException}, exceptions suppressed by the original exception are suppressed by another {@code CheckpointException} or {@code RestoreException}, depends on the throwing method.
```
Intentionally. With reporting the exception centrally there's no reason to inform parent context about an error in a resource in child component.

AntonKozlov replied:
The reply no longer valid. Please keep modifications in the javadoc to minimum as well

rvansa commented:
I've reworded Global Context javadoc, added forgotten javadoc to `recordException`, and replaced calling `semanticContext()` with inlined versions of the `invokeBeforeCheckpoint` and `invokeAfterRestore` methods.

All changes in this PR are related to contexts. I see little point in dosing doc changes individually; API changes are here to support actual implementation of the fix.

AntonKozlov commented on src/java.base/share/classes/jdk/crac/Resource.java:
```diff
@@ -28,23 +28,50 @@
+     * when the checkpoint cannot be performed (e.g. due to this or some other
+     * resource throwing an exception when {@link #beforeCheckpoint(Context)
+     * beforeCheckpoint}.
+     * Therefore, the resource should not have assumptions about it state; it
```
Resource can be sure the beforeCheckpoint was called, and object is exactly in the state at which the beforeCheckpoint has leaved it.

rvansa replied:
I'll reword. The meaning here was to guide user to handling unexpected states, e.g. if the beforeCheckpoint is locking a lock this should not blindly unlock.

AntonKozlov commented on src/java.base/share/classes/jdk/crac/impl/OrderedContext.java:
```diff
@@ -26,28 +26,58 @@
-    public synchronized void register(Resource r) {
-        register(r, order++);
+    public void afterRestore(Context<? extends Resource> context) throws RestoreException {
+        // Note: a resource might attempt to
```
Comment truncated?..

AntonKozlov reviewed (commented):
First itertation, have not looked carefully on the doc and context changes

AntonKozlov commented on src/java.base/share/classes/jdk/crac/impl/AbstractContextImpl.java:
```diff
@@ -22,86 +22,143 @@
+            ce.addSuppressed(t);
+        }
+        if (e.getMessage() != null) {
+            ce.addSuppressed(e);
```
What happens with exceptions suppressed by `e`? will we have the same set of exceptions suppressed by `ce` and `e`?

rvansa replied:
Yes; there's no way to remove already suppressed exceptions, and the message is too valuable to lose it. I could create an exception with the same message but without suppressed exceptions - then I'd lose its stack trace, though.
I don't see much issues in this duplication.

AntonKozlov replied:
It seems we should have allowed the message in the CheckpointException, to avoid this problem. The only place where CheckpointException with the message is called is recursive checkpoint detection from https://github.com/openjdk/crac/pull/6. 

OK for this PR.

AntonKozlov replied:
Hiding all constructors makes sense. I think it will be helpful to have CheckpointException, CheckpointResourceExceptions and descedants something formal, something that can be programmatically queried for the problem. I propose to investigate the possiblity in #64 which touches related pieces.

AntonKozlov reviewed (commented):
Context changes are mostly good, except a small nit

rvansa commented on test/jdk/jdk/crac/ContextOrderTest.java:
```diff
@@ -0,0 +1,340 @@
+    public static void main(String[] args) throws Exception {
+        System.setProperty("java.util.logging.config.file", Utils.TEST_SRC + "/logging.properties");
+
+//        testOrder();
```
Oops, noticed this got commented out.

AntonKozlov reviewed (approved):
LGTM!

Let's wait for tests result

rvansa commented:
/integrate

AntonKozlov commented:
/sponsor

AntonKozlov commented:
I got this after some unrelated modifications (FileDescriptor.beforeCheckpoint uses lambda), and I apparently get auto-deadlock with a single thread involved:

```
"main" #1 prio=5 os_prio=0 cpu=88.95ms elapsed=21.61s tid=0x00007fd670025160 nid=0x27228a in Object.wait()  [0x00007fd6747fd000]
   java.lang.Thread.State: WAITING (on object monitor)
	at java.lang.Object.wait(java.base@17-internal/Native Method)
	- waiting on <0x0000000418002088> (a jdk.internal.crac.JDKContext)
	at java.lang.Object.wait(java.base@17-internal/Object.java:338)
	at jdk.crac.impl.AbstractContextImpl.waitWhileCheckpointIsInProgress(java.base@17-internal/AbstractContextImpl.java:102)
	at jdk.crac.impl.PriorityContext.register(java.base@17-internal/PriorityContext.java:44)
	- locked <0x0000000418002088> (a jdk.internal.crac.JDKContext)
	at jdk.internal.crac.JDKContext.register(java.base@17-internal/JDKContext.java:97)
	at jdk.internal.ref.CleanerImpl$PhantomCleanableRef.<init>(java.base@17-internal/CleanerImpl.java:170)
	at java.lang.ref.Cleaner.register(java.base@17-internal/Cleaner.java:220)
	at java.lang.invoke.MethodHandleNatives$CallSiteContext.make(java.base@17-internal/MethodHandleNatives.java:90)
	at java.lang.invoke.CallSite.<init>(java.base@17-internal/CallSite.java:144)
	at java.lang.invoke.ConstantCallSite.<init>(java.base@17-internal/ConstantCallSite.java:50)
	at java.lang.invoke.InnerClassLambdaMetafactory.buildCallSite(java.base@17-internal/InnerClassLambdaMetafactory.java:270)
	at java.lang.invoke.LambdaMetafactory.metafactory(java.base@17-internal/LambdaMetafactory.java:341)
	at java.lang.invoke.DirectMethodHandle$Holder.invokeStatic(java.base@17-internal/DirectMethodHandle$Holder)
	at java.lang.invoke.Invokers$Holder.invokeExact_MT(java.base@17-internal/Invokers$Holder)
	at java.lang.invoke.BootstrapMethodInvoker.invoke(java.base@17-internal/BootstrapMethodInvoker.java:134)
	at java.lang.invoke.CallSite.makeSite(java.base@17-internal/CallSite.java:315)
	at java.lang.invoke.MethodHandleNatives.linkCallSiteImpl(java.base@17-internal/MethodHandleNatives.java:281)
	at java.lang.invoke.MethodHandleNatives.linkCallSite(java.base@17-internal/MethodHandleNatives.java:271)
	at java.io.FileDescriptor$Resource.beforeCheckpoint(java.base@17-internal/FileDescriptor.java:74)
	at jdk.crac.impl.PriorityContext$SubContext.invokeBeforeCheckpoint(java.base@17-internal/PriorityContext.java:107)
	at jdk.crac.impl.OrderedContext.runBeforeCheckpoint(java.base@17-internal/OrderedContext.java:70)
	at jdk.crac.impl.AbstractContextImpl.beforeCheckpoint(java.base@17-internal/AbstractContextImpl.java:81)
	at jdk.crac.impl.AbstractContextImpl.invokeBeforeCheckpoint(java.base@17-internal/AbstractContextImpl.java:41)
	at jdk.crac.impl.PriorityContext.runBeforeCheckpoint(java.base@17-internal/PriorityContext.java:70)
	at jdk.crac.impl.AbstractContextImpl.beforeCheckpoint(java.base@17-internal/AbstractContextImpl.java:81)
	at jdk.internal.crac.JDKContext.beforeCheckpoint(java.base@17-internal/JDKContext.java:85)
	at jdk.crac.impl.AbstractContextImpl.invokeBeforeCheckpoint(java.base@17-internal/AbstractContextImpl.java:41)
	at jdk.crac.impl.OrderedContext.runBeforeCheckpoint(java.base@17-internal/OrderedContext.java:70)
	at jdk.crac.impl.AbstractContextImpl.beforeCheckpoint(java.base@17-internal/AbstractContextImpl.java:81)
	at jdk.crac.Core.checkpointRestore1(java.base@17-internal/Core.java:116)
	at jdk.crac.Core.checkpointRestore(java.base@17-internal/Core.java:256)
	- locked <0x0000000418002118> (a java.lang.Object)
	at jdk.crac.Core.checkpointRestore(java.base@17-internal/Core.java:241)
	at CheckpointWithOpenFdsTest.exec(CheckpointWithOpenFdsTest.java:49)
	at jdk.test.lib.crac.CracTest.run(CracTest.java:157)
	at jdk.test.lib.crac.CracTest.main(CracTest.java:89)
```

rvansa commented:
Yes, I've ran into this couple of times as well, when invoking lambda or using method reference after CLEANERS priority. The solution is either to remove those dynamically linked call sites, or initialize those call sites eagerly before.

It asks for a general solution, though, adding a priority level for cleaners used only for call sites (a special method would be required in the cleaners). I'll try to sketch that out.

One other idea I had, as I prefer exceptions over deadlocks, is to detect this. Since you don't want to add any methods to the API that would be generally usable by *any* context, it might be still useful to let the C/R code in Core set a static var/thread-local flag in `AbstractContextImpl` and this method would throw rather than block when it detects that it is the same thread. That breaks encapsulation a bit, though - we could have a resource that sets that flag, but it should run before the 'normal' resources, which leads me to the inversion of JDKContext and Global Context and `EARLY` priority I've suggested before.

rvansa commented:
@AntonKozlov Please discuss the general solution in here: https://github.com/openjdk/crac/pull/66

AntonKozlov commented:
Stack trace after merging #66 => the jdk.crac code also cannot use lambdas:

```
"main" #1 prio=5 os_prio=0 cpu=89.78ms elapsed=16.55s tid=0x00007f0a68025160 nid=0x28513b in Object.wait()  [0x00007f0a6d9fd000]
   java.lang.Thread.State: WAITING (on object monitor)
	at java.lang.Object.wait(java.base@17-internal/Native Method)
	- waiting on <0x000000041a802088> (a jdk.internal.crac.JDKContext)
	at java.lang.Object.wait(java.base@17-internal/Object.java:338)
	at jdk.crac.impl.AbstractContextImpl.waitWhileCheckpointIsInProgress(java.base@17-internal/AbstractContextImpl.java:102)
	at jdk.crac.impl.PriorityContext.register(java.base@17-internal/PriorityContext.java:44)
	- locked <0x000000041a802088> (a jdk.internal.crac.JDKContext)
	at jdk.internal.crac.JDKContext.register(java.base@17-internal/JDKContext.java:97)
	at jdk.internal.ref.CleanerImpl$PhantomCleanableRef.<init>(java.base@17-internal/CleanerImpl.java:173)
	at java.lang.invoke.MethodHandleNatives$CallSiteContext.make(java.base@17-internal/MethodHandleNatives.java:93)
	at java.lang.invoke.CallSite.<init>(java.base@17-internal/CallSite.java:144)
	at java.lang.invoke.ConstantCallSite.<init>(java.base@17-internal/ConstantCallSite.java:50)
	at java.lang.invoke.InnerClassLambdaMetafactory.buildCallSite(java.base@17-internal/InnerClassLambdaMetafactory.java:262)
	at java.lang.invoke.LambdaMetafactory.metafactory(java.base@17-internal/LambdaMetafactory.java:341)
	at java.lang.invoke.DirectMethodHandle$Holder.invokeStatic(java.base@17-internal/DirectMethodHandle$Holder)
	at java.lang.invoke.Invokers$Holder.invokeExact_MT(java.base@17-internal/Invokers$Holder)
	at java.lang.invoke.BootstrapMethodInvoker.invoke(java.base@17-internal/BootstrapMethodInvoker.java:134)
	at java.lang.invoke.CallSite.makeSite(java.base@17-internal/CallSite.java:315)
	at java.lang.invoke.MethodHandleNatives.linkCallSiteImpl(java.base@17-internal/MethodHandleNatives.java:285)
	at java.lang.invoke.MethodHandleNatives.linkCallSite(java.base@17-internal/MethodHandleNatives.java:275)
	at jdk.internal.crac.JDKContext.getClaimedFds(java.base@17-internal/JDKContext.java:102)
	at jdk.crac.Core.checkpointRestore1(java.base@17-internal/Core.java:125)
	at jdk.crac.Core.checkpointRestore(java.base@17-internal/Core.java:256)
	- locked <0x000000041a802118> (a java.lang.Object)
	at jdk.crac.Core.checkpointRestore(java.base@17-internal/Core.java:241)
	at CheckpointWithOpenFdsTest.exec(CheckpointWithOpenFdsTest.java:49)
	at jdk.test.lib.crac.CracTest.run(CracTest.java:157)
	at jdk.test.lib.crac.CracTest.main(CracTest.java:89)
```

rvansa commented:
Did you change this method? In #43 this uses loop rather than streams. I think that in the Core code the need to remove lambdas is unavoidable; the scope of what's happening after all notifications and before the `afterRestore` is minimal.

This is similar to early stages of the JDK boot; in some classes you would run into NPE had you tried to use lambdas (e.g. when initializing `JDKContext.PRIORITY_COMPARATOR`).

AntonKozlov commented:
Yes, I did. But in general, these limitations do not look very pleasant for programming. I think we are trying to find a perfect policy for context when there are too many sets of requirements:
- We want some resource to be certainly called, if beforeCheckpoint does a critical clean-up (e.g. wiping security keys), not doing that is a bug. But these don't care about ordering with other resources. In this case, running notification right on registartion looks like an option (as you've proposed in some of the threads), or collecting such late registrations and run all of them, before 
- Some of the resources that do care about ordering does not necessarily care about being certailny called, like those which do optimization clean up like clearing extra cache which does affect correctness.

I'm leaning toward introduction of a few JDKContext variants that will provide different guarantees for being registered resources. Global context still to accomodate them.

GlobalContext:
- PriorityContext
  - ...
- Cleanup // everything that needs clean up, immediate runs of the notification
- ...User Resources...

The scheme actually is not something specific to JDK, user code may have something similar or different, depending on the needs.
